packages/runtime/container-runtime/src/test/pendingStateManager.spec.ts

```diff
@@ -428,6 +458,147 @@ describe("Pending State Manager", () => {
+				})),
+				0,
+			);
+			pendingStateManager.replayPendingStates();
+			assert.strictEqual(resubmittedBatchIds[0], `${clientId}_[0]`);
+			assert.strictEqual(resubmittedBatchIds[1], `${clientId}_[0]`);
+		});
+
+		it("replays pending states with empty batch", () => {
+			// Empty batch is reflected in the pending state manager as a single message
+			// with the following metadata:
+			pendingStateManager.onFlushBatch(
+				[
+					{
+						contents: JSON.stringify({ type: "groupedBatch", contents: [] }),
+						referenceSequenceNumber: 0,
+						metadata: { emptyBatch: true, batchId: "batchId" },
+					},
+				],
+				0,
+			);
+			pendingStateManager.replayPendingStates();
+			assert.strictEqual(resubmittedBatchIds[0], "batchId");
+		});
+	});
+
+	describe("applyStashedOpsAt", () => {
+		it("applyStashedOpsAt", async () => {
+			const applyStashedOps: string[] = [];
+			const messages: IPendingMessage[] = [
+				{
+					type: "message",
+					content: '{"type":"component"}',
+					referenceSequenceNumber: 10,
+					localOpMetadata: undefined,
+					opMetadata: undefined,
+					batchIdContext: { clientId: "CLIENT_ID", batchStartCsn: 1 },
+				},
+				{
+					type: "message",
+					content: '{"type": "component", "contents": {"prop1": "value"}}',
+					referenceSequenceNumber: 11,
+					localOpMetadata: undefined,
+					opMetadata: undefined,
+					batchIdContext: { clientId: "CLIENT_ID", batchStartCsn: 2 },
+				},
+			];
+
+			const pendingStateManager = new PendingStateManager(
+				{
+					applyStashedOp: async (content) => applyStashedOps.push(content),
+					clientId: () => "clientId",
+					close: () => {},
+					connected: () => true,
+					reSubmitBatch: () => {},
+					isActiveConnection: () => false,
+					isAttached: () => true,
+				},
+				{ pendingStates: messages },
+				logger,
+			);
+			await pendingStateManager.applyStashedOpsAt();
+			assert.strictEqual(applyStashedOps.length, 2);
+			assert.strictEqual(pendingStateManager.pendingMessagesCount, 2);
+		});
+
+		it("applyStashedOpsAt for empty batch", async () => {
+			const applyStashedOps: string[] = [];
+			const messages: IPendingMessage[] = [
+				{
+					type: "message",
+					content: '{"type":"groupedBatch", "contents": []}',
+					referenceSequenceNumber: 10,
+					opMetadata: undefined,
+					localOpMetadata: undefined,
+					batchIdContext: { clientId: "CLIENT_ID", batchStartCsn: 1 },
+				},
+			];
+
+			const pendingStateManager = new PendingStateManager(
+				{
+					applyStashedOp: async (content) => applyStashedOps.push(content),
+					clientId: () => "clientId",
+					close: () => {},
+					connected: () => true,
+					reSubmitBatch: () => {},
+					isActiveConnection: () => false,
+					isAttached: () => true,
+				},
+				{ pendingStates: messages },
+				logger,
+			);
+			await pendingStateManager.applyStashedOpsAt();
+			assert.strictEqual(applyStashedOps.length, 0);
+			assert.strictEqual(pendingStateManager.pendingMessagesCount, 1);
```

Comment:
You could also assert that the pending message is the `emptyBatch` localOpMetadata